TLM/Benchmarks/TaskListPerfTests.cs

```diff
@@ -0,0 +1,212 @@
+using BenchmarkDotNet.Attributes;
+using System;
+using System.Collections.Generic;
+using TrafficManager.Util;
+
+namespace Benchmarks {
+    public class TaskListPerfTests {
+
+        [GlobalSetup]
+        public void Setup() {
+            TestData.InitialiseLists();
+            TestData.GenerateDeltas();
+        }
+
+        [Benchmark]
+        public void TaskList_SimulatedCamMoves() {
+
+            var list = TestData.testTaskList;
+
+            foreach (var delta in TestData.Deltas) {
+
+                // update & remove
+                for (int index = 0; index < list.Size; index++) {
+                    if (delta.Retained.Contains(list.Tasks[index].ID)) {
+                        // update code would be here
+                    } else {
+                        list.RemoveAt(index);
+                    }
+                }
+
+                // add new tasks
+                for (int i = 0; i < delta.Added.Length; i++) {
+                    list.Add(new TestStruct { ID = delta.Added[i] });
+                }
+
+            }
+        }
+
+        [Benchmark]
+        public void TaskList_Optimised_SimulatedCamMoves() {
+
+            var list = TestData.testTaskListOptimised;
+
+            foreach (var delta in TestData.Deltas) {
+
+                // update & remove
+                for (int index = 0; index < list.Size; index++) {
+                    if (delta.Retained.Contains(list.Tasks[index].ID)) {
+                        // update code would be here
+                    } else {
+                        list.RemoveAt(index);
+                    }
+                }
+
+                list.PrepareToAdd(delta.Added.Length);
+
+                // add new tasks
+                for (int i = 0; i < delta.Added.Length; i++) {
+                    ref var task = ref list.Tasks[list.NextUsableTaskIndex];
+
+                    task.ID = delta.Added[i];
+                    task.IsActive = true;
+                }
+
+            }
+        }
+
+        [Benchmark]
+        public void FastList_SimulatedCamMoves() {
+
+            var list = TestData.testFastList;
+
+            foreach (var delta in TestData.Deltas) {
+
+                // update & remove
+                for (int index = list.m_size; index-- > 0;) {
+                    if (delta.Retained.Contains(list.m_buffer[index].ID)) {
+                        // update code would be here
+                    } else {
+                        list.RemoveAt(index);
+                    }
+                }
+
+                // add new tasks
+                for (int i = 0; i < delta.Added.Length; i++) {
+                    list.Add(new TestStruct { ID = delta.Added[i], IsActive = true, });
```

Comment:
> `int[]` index access in a loop here takes 50% of whole benchmark time. When you replace `ID = delta.Added[i]` with `ID = i` benchmark will be 2x as fast.

The ID would actually be an `InstanceID` of the segment / node / building / etc. So I need to store that ID so that next time the cam moves I can use the `Retained` hashset to work out what to keep (reposition) and what to delete. If there's a faster/better way to handle that lmk.

> Maybe split into multiple benchmarks showing the impact of number of elements, add more scenarios like e.g.:
> 
>     * add a lot of new items,
>     * remove a lot of items,

Yup, I can split them in to Modify (update/delete) and Add sections.

>     * completely replace whole collection

If the `Retained` hashset is empty a simple `Clear()` will be much faster so I'll add that in.

>       That may show us how it really behaves in certain situations - If Deltas contain multiple scenarios I would like to see how each one behaves in isolation then some/all combined to give general look about potential and limitations of what is it capable and when it shouldn't ever be used if tests show issues.

I'll try and split it out to benchmark some possible scenarios.

>       Benchmarks could also help finding good enough initial size of arrays/collections.

This will depend a lot on what the camera is looking at, how many overlays are active, and how much culling we do (eg. based on distance between instance and camera) prior to lists of IDs ever reaching this code.

Reply:
These sort of stats better?

|                                                 Method |     Mean |     Error |   StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------------------------------- |---------:|----------:|---------:|------:|------:|------:|----------:|
|          TaskList_01_Add6kItems_To_UnpreparedEmptyList | 35.88 μs | 18.278 μs | 1.002 μs |     - |     - |     - |         - |
|          FastList_01_Add6kItems_To_UnpreparedEmptyList | 56.56 μs | 20.722 μs | 1.136 μs |     - |     - |     - |         - |
|            TaskList_02_Add6kItems_To_PreparedEmptyList | 36.17 μs | 49.451 μs | 2.711 μs |     - |     - |     - |         - |
|            FastList_02_Add6kItems_To_PreparedEmptyList | 33.20 μs | 40.967 μs | 2.246 μs |     - |     - |     - |         - |
|    TaskList_03_ManuallyAdd6kItems_To_PreparedEmptyList | 24.45 μs |  4.207 μs | 0.231 μs |     - |     - |     - |         - |
|    FastList_03_ManuallyAdd6kItems_To_PreparedEmptyList | 23.98 μs |  9.170 μs | 0.503 μs |     - |     - |     - |         - |
| TaskList_04_AltManuallyAdd6kItems_To_PreparedEmptyList | 24.38 μs |  9.618 μs | 0.527 μs |     - |     - |     - |         - |